### PR TITLE
Validations BLI 445

### PR DESCRIPTION
**Pull request by allaVolkov**
BLI 445 validations related to unsupported features

### Checklist
- [x] Code compiles correctly
- [x] Relevant tests were added (unit / contract / integration)
- [x] Relevant logs were added
- [x] Formating and linting run locally successfully
- [x] All tests passing
- [ ] UA review
- [ ] Design is documented
- [ ] Extended the README / documentation, if necessary
- [ ] Open source is approved

### CODE REVIEW

**Comment by tal-sapan** [comment]
Run `make format && make lint` before pushing so the build won't fail

File: validations/builder_schema.go [hunk 1]
```diff
@@ -1,42 +1,52 @@
...
 	"gopkg.in/yaml.v3"
 
 	"github.com/SAP/cloud-mta/mta"
 )
 
+func checkStringParam(params map[string]interface{}, parentNode *yaml.Node, paramsListName string, paramName string) []YamlValidationIssue {
+	var issues []YamlValidationIssue
+
+	if params != nil && params[paramName] != nil {
+		paramsNode := getPropValueByName(parentNode, paramsListName)
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/builder_schema.go]: moved

File: validations/builder_schema.go [hunk 2]
```diff
@@ -1,42 +1,52 @@
 package validate
 
 import (
+	"fmt"
 	"gopkg.in/yaml.v3"
 
 	"github.com/SAP/cloud-mta/mta"
 )
 
+func checkStringParam(params map[string]interface{}, parentNode *yaml.Node, paramsListName string, paramName string) []YamlValidationIssue {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/builder_schema.go]: renamed

[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/builder_schema.go]: done

File: validations/builder_schema.go [hunk 3]
```diff
@@ -1,42 +1,52 @@
...
 
+	issues = append(issues, checkStringParam(mta.Parameters, mtaNode, parametersYamlField, deployModeYamlField)...)
+
 	modulesNode := getPropContent(mtaNode, modulesYamlField)
 
 	for i, module := range mta.Modules {
-		if module.BuildParams != nil && module.BuildParams[builderYamlField] != nil {
-			_, ok := module.BuildParams[builderYamlField].(string)
+		issues = append(issues, checkStringParam(module.BuildParams, modulesNode[i], buildParametersYamlField, builderYamlField)...)
+		if module.BuildParams != nil && module.BuildParams[commandsYamlField] != nil {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/builder_schema.go]: done

File: validations/builder_schema.go [hunk 4]
```diff
@@ -1,42 +1,52 @@
...
+			paramNode := getPropValueByName(paramsNode, paramName)
+			issues = appendIssue(issues,
+				fmt.Sprintf(`the "%s" property is defined incorrectly; the property must be a string`, paramName),
+				paramNode.Line)
+		}
+	}
+	return issues
+}
+
 func checkBuilderSchema(mta *mta.MTA, mtaNode *yaml.Node, source string) []YamlValidationIssue {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/builder_schema.go]: we have to understand what's going on with mta schema definition, if build-parameters will be defined as structure and not as map[string]interface{}. If yes we'll adjust and simplify the code accordingly.
If not we have to make refactoring and move all the logic around build-parameters to mt-build-tool and define corresponding schema there.

File: validations/semantic_deployer_constraints.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+		// case !ok is handled by schema validations
+		if ok && deployMode == html5RepoDeployMode {
+			issues = append(issues, checkHTMLModulesParams(mta, mtaNode)...)
+		}
+	}
+
+	return issues, nil
+}
+
+func checkHTMLModulesParams(mta *mta.MTA, mtaNode *yaml.Node) []YamlValidationIssue {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/semantic_deployer_constraints.go]: done

File: validations/semantic_deployer_constraints.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+	for i, module := range mta.Modules {
+		if module.Type == html5ModuleType {
+			issues = append(issues, checkHTMLModuleParams(module, modulesNode[i])...)
+		}
+	}
+
+	return issues
+}
+
+func checkHTMLModuleParams(module *mta.Module, moduleNode *yaml.Node) []YamlValidationIssue {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/semantic_deployer_constraints.go]: done

File: validations/semantic_deployer_constraints.go [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+	supportedPlatformsDefined := false
+	buildResultDefined := false
+	if module.BuildParams != nil {
+		supportedPlatformsDefined = module.BuildParams[supportedPlatformsYamlField] != nil
+		buildResultDefined = module.BuildParams[buildResultYamlField] != nil
+	}
+
+	if !supportedPlatformsDefined && !buildResultDefined {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/semantic_deployer_constraints.go]: as discussed with Natalie that is the check to be done. If user put not expected values it's his responsibility.

File: validations/semantic_deployer_constraints_test.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/SAP/cloud-mta/mta"
+)
+
+var _ = Describe("SemanticDeployerConstraints", func() {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/semantic_deployer_constraints_test.go]: we are not expecting values, we just check that fields are provided. (by Natalia)

File: validations/semantic_deployer_constraints_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+		checkDeployerConstrError(errors[0], missingConfigsMsg, 10, "ui5app1")
+		checkDeployerConstrError(errors[1], missingConfigsMsg, 13, "ui5app2")
+		checkDeployerConstrError(errors[2], missingConfigMsg, 18, "ui5app3", buildResultYamlField, )
+		checkDeployerConstrError(errors[3], missingConfigMsg, 23, "ui5app4", supportedPlatformsYamlField)
+	})
+})
+
+func checkDeployerConstrError(issue YamlValidationIssue, message string, line int, params ...string) {
+	if len(params) == 1 {
+		Ω(issue.Msg).Should(Equal(fmt.Sprintf(message, params[0], missingConfigDocLink)))
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/semantic_deployer_constraints_test.go]: didn't succeed

File: validations/semantic_deployer_constraints_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (tal-sapan) [reply validations/semantic_deployer_constraints_test.go]: So the expected behavior is that no error is returned. It should still be tested.

File: validations/builder_schema.go [hunk 10]
```diff
@@ -7,45 +7,46 @@ import (
...
 	}
-	return issues
+	return nil
 }
 
 func checkBuilderSchema(mta *mta.MTA, mtaNode *yaml.Node, source string) []YamlValidationIssue {
 	var issues []YamlValidationIssue
 
-	issues = append(issues, checkStringParam(mta.Parameters, mtaNode, parametersYamlField, deployModeYamlField)...)
+	issues = append(issues, checkStringProperty(mta.Parameters, getPropValueByName(mtaNode, parametersYamlField), deployModeYamlField)...)
```
Comment (tal-sapan) [inline validations/builder_schema.go]: Won't this add nil to the issues slice if there's no error?

File: validations/semantic_deployer_constraints.go [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+	"fmt"
+	"gopkg.in/yaml.v3"
+
+	"github.com/SAP/cloud-mta/mta"
+)
+
+const (
+	missingConfigDocLink = "https://sap.github.io/cloud-mta-build-tool/migration/#features-that-are-handled-differently-in-the-cloud-mta-build-tool"
+	missingConfigsMsg    = `the "%s" module does not contain the mandatory "supported-platforms" and "build-result" build configurations; see %q`
+	missingConfigMsg     = `the "%s" module does not contain the mandatory the mandatory "%s" build configuration; see %q`
```
[the remark this answers is not in the document]
Comment (tal-sapan) [reply validations/semantic_deployer_constraints.go]: This is still relevant...

**Review by tal-sapan — COMMENTED** [review]
See comments

File: validations/builder_schema.go (hunk 10 above)
Comment (allaVolkov) [reply validations/builder_schema.go]: no, it receives slice, that can be empty or nil. "a nil slice is functionally equivalent to a zero-length slice, even though it points to nothing"  (from: https://blog.golang.org/slices). It works in both directions: we can append to nil slice and we can append a nil slice.

File: validations/builder_schema.go [hunk 12]
```diff
@@ -7,45 +7,46 @@ import (
...
-	if params != nil && params[paramName] != nil {
-		paramsNode := getPropValueByName(parentNode, paramsListName)
-		_, ok := params[paramName].(string)
-		if !ok {
-			paramNode := getPropValueByName(paramsNode, paramName)
-			issues = appendIssue(issues,
-				fmt.Sprintf(`the "%s" property is defined incorrectly; the property must be a string`, paramName),
-				paramNode.Line)
+	_, ok := props[propName].(string)
+	if props != nil && props[propName] != nil && !ok {
```
[the remark this answers is not in the document]
Comment (allaVolkov) [reply validations/builder_schema.go]: done

**Review by tal-sapan — APPROVED** [review]

**Pull request merged**